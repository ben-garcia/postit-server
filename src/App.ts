import { ApolloServer } from 'apollo-server-express';
import bcrypt from 'bcrypt';
import cookieParser from 'cookie-parser';
import dotenv from 'dotenv';
import express, { Application } from 'express';
import helmet from 'helmet';
import jwt from 'jsonwebtoken';
import { Container } from 'typedi';
import { getRepository } from 'typeorm';

import {
  Community,
  GeneralPreferences,
  NotificationPreferences,
  EmailNotificationPreferences,
  Profile,
  User,
} from './entities';
import {
  createEmailTemplate,
  createRedisClient,
  createTransporter,
  createSchema,
  formatResponse,
} from './utils';

dotenv.config();

class App {
  public app: Application;
  private server: ApolloServer;

  constructor() {
    this.app = express();

    this.initializeMiddleware();
    this.initializeServer();
  }

  private async initializeMiddleware() {
    // Set values on the injected properties.
    Container.set('userRepository', getRepository(User));
    Container.set('profileRepository', getRepository(Profile));
    Container.set(
      'generalPreferencesRepository',
      getRepository(GeneralPreferences)
    );
    Container.set(
      'notificationPreferencesRepository',
      getRepository(NotificationPreferences)
    );
    Container.set(
      'emailNotificationPreferencesRepository',
      getRepository(EmailNotificationPreferences)
    );
    Container.set('communityRepository', getRepository(Community));
    Container.set('transporter', await createTransporter());
    Container.set('bcrypt', bcrypt);
    Container.set('jwt', jwt);
    Container.set('redisClient', createRedisClient());
    Container.set('emailTemplate', createEmailTemplate());

    const cookieSecret = process.env.COOKIE_SECRET || 'cookiesecret';

    this.app.use(cookieParser(cookieSecret));
    this.app.use(
      helmet({
        // play nice with graphql playground
        contentSecurityPolicy:
          process.env.NODE_ENV === 'production' ? undefined : false,
      })
    );

    const cors = {
      // send relevant cookies
      credentials: true,
      methods: ['POST'],
      // should match the origin of the request
      origin: process.env.CLIENT_URL ?? 'http://localhost:3000',
    };

    this.server.applyMiddleware({
      app: this.app,
      cors,
    });
  }

  private async initializeServer() {
    // builds the GraphQL schema using the resolver classes.
    const schema = await createSchema();

    this.server = new ApolloServer({
      context: ({ req, res }) => ({ req, res }),
      // @ts-ignore
      formatResponse,
      schema,
    });
  }

  public async listen() {
    this.app.listen(4000, () =>
      // eslint-disable-next-line no-console
      console.log(
        `ðŸš€ Server ready at http://localhost:4000${this.server.graphqlPath}`
      )
    );
  }
}

export default App;
