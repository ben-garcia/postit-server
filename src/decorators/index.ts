import { IsEmailUnique } from './isEmailUnique';
import { IsUsernameUnique } from './isUsernameUnique';

export { IsEmailUnique, IsUsernameUnique };
