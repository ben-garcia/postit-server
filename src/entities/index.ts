import GeneralPreferences from './GeneralPreferences';
import Profile from './Profile';
import User from './User';

export { GeneralPreferences, Profile, User };
