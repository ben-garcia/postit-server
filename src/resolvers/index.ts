import AuthResolver from './AuthResolver';
import UserResolver from './UserResolver';

export { AuthResolver, UserResolver };
