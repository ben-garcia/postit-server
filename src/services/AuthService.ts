import * as JWT from 'jsonwebtoken';
import { Inject, Service } from 'typedi';

interface Payload {
  email: string;
  username: string;
}

/**
 * This Service is responsible for:
 *
 * creating jwt access and refresh tokens.
 * revoking refresh token.
 */
@Service()
class AuthService {
  private jwt: typeof JWT;

  constructor(@Inject('jwt') jwt: typeof JWT) {
    this.jwt = jwt;
  }

  /**
   * Create access token using payload that expires in 15 minutes.
   *
   * Create refresh token using access token that expires in 1 year.
   */
  createTokens(payload: Payload): [string, string] {
    const jwtSecret = process.env.JWT_SECRET || 'thisisasecret';
    const jwtRefreshSecret = process.env.JWT_REFRESH_SECRET || 'secretrefresh';
    const accessToken = this.jwt.sign(payload, jwtSecret, {
      expiresIn: '15m',
    });
    const refreshToken = this.jwt.sign(
      { ...payload, accessToken },
      jwtRefreshSecret,
      {
        expiresIn: '1y',
      }
    );

    return [accessToken, refreshToken];
  }
}

export default AuthService;
