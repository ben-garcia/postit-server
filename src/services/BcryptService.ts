import * as Bcrypt from 'bcrypt';
import { Inject, Service } from 'typedi';

/**
 * This Service is responsible for:
 *
 * creating comparing the hashed password in the db
 * with the password in the request when attempting to log in.
 */
@Service()
class BcryptService {
  private bcrypt: typeof Bcrypt;

  constructor(@Inject('bcrypt') bcrypt: typeof Bcrypt) {
    this.bcrypt = bcrypt;
  }

  /**
   * Verify if the password match
   *
   * @param passwordToCheck the password in the request
   * @param userPassword the hashed password associated with the user
   */
  async validatePassword(
    passwordToCheck: string,
    userPassword: string
  ): Promise<boolean> {
    return this.bcrypt.compare(passwordToCheck, userPassword);
  }
}

export default BcryptService;
