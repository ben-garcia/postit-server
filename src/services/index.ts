export { default as BcryptService } from './BcryptService';
export { default as CommunityService } from './CommunityService';
export { default as GeneralPreferencesService } from './GeneralPreferencesService';
export { default as NotificationPreferencesService } from './NotificationPreferencesService';
export { default as EmailNotificationPreferencesService } from './EmailNotificationPreferencesService';
export { default as JwtService } from './JwtService';
export { default as MailService } from './MailService';
export { default as ProfileService } from './ProfileService';
export { default as RedisService } from './RedisService';
export { default as UserService } from './UserService';
