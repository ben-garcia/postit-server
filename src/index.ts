import 'reflect-metadata';
import { ApolloServer } from 'apollo-server-express';
import cookieParser from 'cookie-parser';
import dotenv from 'dotenv';
import express from 'express';
import jwt from 'jsonwebtoken';
import { Container } from 'typedi';
import { createConnection, getRepository } from 'typeorm';

import { User } from './entities';
import { createTransporter, createSchema } from './utils';

dotenv.config();

(async () => {
  // establish the connection to the database.
  await createConnection();

  // Set values on the injected properties.
  Container.set('userRepository', getRepository(User));
  Container.set('transporter', await createTransporter());
  Container.set('jwt', jwt);

  const app = express();
  const cookieSecret = process.env.COOKIE_SECRET || 'cookiesecret';

  app.use(cookieParser(cookieSecret));
  // builds the GraphQL schema using the resolver classes.
  const schema = await createSchema();
  const server = new ApolloServer({
    context: ({ req, res }) => ({ req, res }),
    schema,
  });

  server.applyMiddleware({ app });

  app.listen(4000, () =>
    // eslint-disable-next-line no-console
    console.log(`ðŸš€ Server ready at http://localhost:4000${server.graphqlPath}`)
  );
})();
