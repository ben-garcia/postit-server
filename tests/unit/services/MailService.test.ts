import { MailService } from '../../../src/services';

describe('MailService', () => {
  let mailService: MailService;
  let mockTransporter: any;
  let mockEmailTemplate: any;
  const email = 'bar@example.com';
  const username = 'bar';
  const subject = 'Verify your Postit email address';
  const base64String = 'jksjlfjlaksjkslajfksafj';
  const text = `Postit		u/${username} [http://localhost:3000/user/${username}]

Hi there,

Your email address ${email} has been added to your
${username} Postit account. But wait, we’re not done yet...

To finish verifying your email address and securing your account, click the
button below.

Verify Email Address
[http://localhost:3000/verification/${base64String}]
`;
  const html = `<!DOCTYPE html>
<html>
	<head>
		<title>Email Verification</title>
		
	</head>
	<body>
		<div class="container" style="height: 100%; width: 100%;">
			<div class="container__inner" style="font-family: Helvetica, Arial, sans-serif; line-height: 18px; margin: 0 auto; width: 45%;">
				<div class="top-nav" style="display: flex; justify-content: space-between;">
					<span class="text-lg" style="font-size: 30px;">postit</span>	
					<a class="no-underline" href="http://localhost:3000/user/${username}" style="text-decoration: none;">
						<span class="color-gray text-sm" style="font-size: 10px; color: #7a9299;">
							u/${username}
						</span>
					</a>
				</div>
				<br>
				<br>
				<div class="color-black" style="color: #000;">
					<p>Hi there,</p>
					<p>
						Your email address ${email} has been added to your
						<br>
						${username} Postit account. But wait, we’re not done yet...
					</p>
					<p>
						To finish verifying your email address and securing your account, click the
						<br>
						button below.
					</p>
				</div>
				<div class="button-container" style="display: flex; justify-content: center; margin-top: 30px;">
					<a href="http://localhost:3000/verification/jksjlfjlaksjkslajfksafj" class="button no-underline" style="text-decoration: none; background: #0079d3; border-radius: 4px; font-size: 12px; font-weight: 700; padding: 8px; text-align: center;">
						<span class="button__text color-white" style="color: #fff; padding: 0 22px;">
							Verify Email Address
						</span>
					</a>
			</div>
		</div>
	</div></body>
</html>

`;

  beforeAll(() => {
    mockEmailTemplate = {
      renderAll: jest.fn().mockImplementation(() => ({
        html,
        subject,
        text,
      })),
    };

    mockTransporter = {
      sendMail: jest.fn().mockReturnValue({
        response: {
          length: jest.fn(),
          lastIndexOf: jest.fn(),
          substring: jest.fn(),
        },
      }),
    };

    mailService = new MailService(mockEmailTemplate, mockTransporter);
  });

  it('should be a module', () => {
    expect(mailService).toBeDefined();
  });

  describe('sendVerificationEmail', () => {
    it('should call the transporter.sendMail and emailTemplate.renderAll', async () => {
      const sendMailExpectedArgs = {
        from: 'Postit <support@postit.com>',
        html,
        subject: 'Verify your Postit email address',
        to: email,
        text,
      };
      const clientUrl = 'http://localhost:3000';
      const renderAllExpectedArgs = {
        base64String,
        clientUrl,
        email,
        username,
      };

      await mailService.sendVerificationEmail(email, username, base64String);

      expect(mockTransporter.sendMail).toHaveBeenCalledTimes(1);
      expect(mockTransporter.sendMail).toHaveBeenCalledWith(
        sendMailExpectedArgs
      );

      expect(mockEmailTemplate.renderAll).toHaveBeenCalledTimes(1);
      expect(mockEmailTemplate.renderAll).toHaveBeenCalledWith(
        'verification',
        renderAllExpectedArgs
      );
    });

    it('should match text snapshot', () => {
      expect(text).toMatchSnapshot();
    });

    it('should match html snapshot', () => {
      expect(html).toMatchSnapshot();
    });
  });
});
