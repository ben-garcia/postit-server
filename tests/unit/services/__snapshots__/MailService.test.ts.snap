// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MailService sendVerificationEmail should match html snapshot 1`] = `
"<!DOCTYPE html>
<html>
	<head>
		<title>Email Verification</title>
		
	</head>
	<body>
		<div class=\\"container\\" style=\\"height: 100%; width: 100%;\\">
			<div class=\\"container__inner\\" style=\\"font-family: Helvetica, Arial, sans-serif; line-height: 18px; margin: 0 auto; width: 45%;\\">
				<div class=\\"top-nav\\" style=\\"display: flex; justify-content: space-between;\\">
					<span class=\\"text-lg\\" style=\\"font-size: 30px;\\">postit</span>	
					<a class=\\"no-underline\\" href=\\"http://localhost:3000/user/bar\\" style=\\"text-decoration: none;\\">
						<span class=\\"color-gray text-sm\\" style=\\"font-size: 10px; color: #7a9299;\\">
							u/bar
						</span>
					</a>
				</div>
				<br>
				<br>
				<div class=\\"color-black\\" style=\\"color: #000;\\">
					<p>Hi there,</p>
					<p>
						Your email address bar@example.com has been added to your
						<br>
						bar Postit account. But wait, we’re not done yet...
					</p>
					<p>
						To finish verifying your email address and securing your account, click the
						<br>
						button below.
					</p>
				</div>
				<div class=\\"button-container\\" style=\\"display: flex; justify-content: center; margin-top: 30px;\\">
					<a href=\\"http://localhost:3000/verification/jksjlfjlaksjkslajfksafj\\" class=\\"button no-underline\\" style=\\"text-decoration: none; background: #0079d3; border-radius: 4px; font-size: 12px; font-weight: 700; padding: 8px; text-align: center;\\">
						<span class=\\"button__text color-white\\" style=\\"color: #fff; padding: 0 22px;\\">
							Verify Email Address
						</span>
					</a>
			</div>
		</div>
	</div></body>
</html>

"
`;

exports[`MailService sendVerificationEmail should match text snapshot 1`] = `
"Postit		u/bar [http://localhost:3000/user/bar]

Hi there,

Your email address bar@example.com has been added to your
bar Postit account. But wait, we’re not done yet...

To finish verifying your email address and securing your account, click the
button below.

Verify Email Address
[http://localhost:3000/verification/jksjlfjlaksjkslajfksafj]
"
`;
